### PR TITLE
Session

### LINKED ISSUE #60: crash on large context sizes
For prompts larger than ~100 tokens the server will sometimes crash

### LINKED ISSUE #108: Error after prompting Llama2 on M1.
After starting Ollama and running Llama2, any prompt results in:
`Error: Post "http://127.0.0.1:11434/api/generate": EOF
`
Installed on M1 Macbook Pro
Ollama reports that it is running.
'Ollama list' reports llama2:latest.

Is this a memory issue?

### PR DESCRIPTION
**Pull request by mxyng**
an active model is kept in memory until another session is requested or the session has expired, freeing any memory associated with the model.

This adds a `SessionDuration` field to the generate request to customize the session window (default 5m) and a `SessionExpiresAt` field to the generate response informing users of when the session will expire. 

* A session duration value of `-1` disables session expiration. 
* A session duration value of `0` disables model caching, i.e. models will be garbage collected as soon as generation is complete

resolves #60 
resolves #108

### CODE REVIEW

**Comment by pdevine** [comment]
This started going bezerk while using it after about 3-4 queries. Not sure if it's running out of memory here or not, or somehow the context is getting whacked?

**Comment by mxyng** [comment]
This should be ready for review

**Comment by mxyng** [comment]
That warning will disappear once the branch is rebased on main

**Comment by BruceMacD** [comment]
with this change ollama no longer panics on exit while running a model, but a subsequent request hangs (no response is ever sent as far as I can tell).

Steps to reproduce:
1. Start an interactive session: `ollama run llama2`
2. Send a query
3. `ctrl+c` out while the response is being generated 
4. Start another interactive session: `ollama run llama2`
5. Send a query
6. Interactive session hangs forever and no response is received

File: llama/llama.go [hunk 1]
```diff
@@ -82,29 +82,37 @@ import (
...
 	params *C.struct_llama_context_params
 	model  *C.struct_llama_model
 	ctx    *C.struct_llama_context
 
+	lastN  deque[C.llama_token]
+	embd   []C.llama_token
+	cursor int
+
+	mu sync.Mutex
+	gc bool
```
Comment (BruceMacD) [inline llama/llama.go:105]: This is the parameter used to stop a loaded model?

File: server/routes.go [hunk 2]
```diff
@@ -42,38 +47,79 @@ func generate(c *gin.Context) {
...
-		return
+	if activeSession.expireTimer != nil {
+		activeSession.expireTimer.Stop()
 	}
 
-	llm, err := llama.New(model.ModelPath, opts)
+	sessionDuration := req.SessionDuration
+	sessionID := activeSession.id
+
+	activeSession.expireAt = time.Now().Add(sessionDuration.Duration)
```
Comment (BruceMacD) [inline server/routes.go:84]: Does this cause a connecton to close after a set amount of time?

File: llama/llama.go (hunk 1 above)
Comment (mxyng) [reply llama/llama.go:105]: Yes it's to signal the model to terminate prediction

File: server/routes.go (hunk 2 above)
Comment (mxyng) [reply server/routes.go:84]: Yes if it hasn't been used recently. Recently is determined by session duration. The default value is 5 minutes. Reusing the same session will terminate the previous timer and create a new timer ensuring the session is only terminated on inactivity

**Comment by mxyng** [comment]
> with this change ollama no longer panics on exit while running a model, but a subsequent request hangs (no response is ever sent as far as I can tell).

This is likely due to the mutex lock. The generate is wrapped in a lock to ensure a generation performs optimally. One downside is subsequent requests will wait on the previous request which may take a long time, even if the client cancels. This is somewhat mitigated by a GC check before generating the next word. However, the initial prompt evaluation may be significant and the request handler is locked for a long time. This is probably what you're seeing.

One solution to this is the decrease the batch size of evals which should decrease the max time for eval to return.

What are you inputs so I can more accurately reproduce?

**Comment by BruceMacD** [comment]
The issue of the prompt getting stuck on cancel has been fixed as of the recent changes.

**Review by BruceMacD — APPROVED** [review]
this looks good to go

**Comment by pdevine** [comment]
@mxyng were there updates to this? I was still seeing issues as of Friday

File: llama/llama.go [hunk 3]
```diff
@@ -144,39 +154,130 @@ func New(model string, opts api.Options) (*llama, error) {
...
+		llm.NumKeep = len(tokens)
+	}
+
+	// min(llm.NumCtx - 4, llm.NumKeep)
+	if llm.NumCtx-4 < llm.NumKeep {
+		llm.NumKeep = llm.NumCtx - 4
+	}
+
+	if len(tokens) >= llm.NumCtx {
+		// truncate input
```
Comment (BruceMacD) [inline llama/llama.go:234]: doesn't have to be part of this PR, but I'd like some kind of warning when my input gets truncated. For example if I pass in a large document to be summarized I would like to know that the model hasn't actually received the full context.

File: llama/llama.go [hunk 4]
```diff
@@ -144,39 +154,130 @@ func New(model string, opts api.Options) (*llama, error) {
...
+			lastN = append(lastN, int(t))
+		}
+
+		return true
+	})
+
+	timings := C.llama_get_timings(llm.ctx)
+	fn(api.GenerateResponse{
+		Done:               true,
+		Context:            lastN,
```
Comment (BruceMacD) [inline llama/llama.go:209]: What is this context? The last token index which was generated? It doesnt seem like its used for any logic on the server at the moment.

Comment (mxyng) [reply llama/llama.go]: This is the conversation up until this point according to the model runner. It's returned to the caller as raw tokens which passes it back when sending a new request. The server then reconstructs the full context based on the input prompt and context values.

File: llama/utils.go [hunk 5]
```diff
@@ -1,104 +1,15 @@
...
+	"time"
+)
 
-func (d *deque[T]) Data() (data []T) {
-	for n := d.head; n != nil; n = n.next {
-		data = append(data, n.t)
+func parseDurationMs(ms float64) time.Duration {
+	dur, err := time.ParseDuration(fmt.Sprintf("%fms", ms))
+	if err != nil {
+		panic(err)
```
Comment (BruceMacD) [inline llama/utils.go:11]: we shouldn't cause the whole server to halt in the case we can't parse this, even though its unexpected

**Review by BruceMacD — APPROVED** [review]
this has been stable in my testing, major improvement

**Comment by jmorganca** [comment]
Looks great! LGTM

**Review by jmorganca — APPROVED** [review]

**Review by pdevine — APPROVED** [review]

**Pull request merged**